### ISSUE TITLE
Rolling updates should fail early if a pod fails to start

### ISSUE THREAD

**Issue by jsravn** [kubernetes/kubernetes#18568]
In 1.0.x, rolling update didn't block, so we could check things like a failed pod while waiting for the update to complete.

In 1.1.x, rolling update does block up to the timeout amount. This is great, however it waits the entire timeout even if the update fails immediately. For instance, if you specify a non-existent image the pod will report a `PullImageError`. At that point the rolling update could abort immediately. Instead it blocks for the 5 minute timeout. When frequently developing and deploying this 5 minute timeout becomes painful.

I suggest rolling-update fails immediately if any of the new pods are not in a 'Pending' or 'Running' state. Not sure if there is a set of non-error states we can use.

**Comment by jsravn** [comment]
I was trying to see if I could put together a poc for this, but it doesn't look like there is a programmatic way to check for a failed pod deploy (as of 1.1 codebase at least). It is always in `Pending` state. The only indicator of a problem is in the container state reason string field, but this can be almost anything depending on the container manager (and has changed from 1.1 to master for the pull image error on docker) - don't think I can rely on this.

It would work for `Terminated` state but I believe that only happens if the application in the container errors. Everything else (pulling image, resource limits?) gets stuck in a `Pending` state. It feels like we need another state such as `FailedDeploy`. Or lump it into `Terminated`, since it's not pending if the error is unrecoverable.

**Comment by jsravn** [comment]
The more I think about it, the more I think it's a bug in kubelet with docker. I think it should report the pod as `Failed` if the image pull fails. Then we can have some sensible policies to deal with the failure (whether it's to abort the rolling update, or to try pulling the image again).

**Comment by kargakis** [comment]
> The more I think about it, the more I think it's a bug in kubelet with docker. I think it should report the pod as Failed if the image pull fails. Then we can have some sensible policies to deal with the failure (whether it's to abort the rolling update, or to try pulling the image again).

+1

cc: @vishh @yujuhong

**Comment by vishh** [comment]
I don't think the container can be marked as `Failed` because it is possible for a retry of an image pull to succeed. I do agree that we need to provide some means to determine if a container is failing to start because of image related issues. cc @bgrant0607

**Comment by bgrant0607** [comment]
cc @ironcladlou @nikhiljindal @janetkuo @kargakis

**Comment by yujuhong** [comment]
I agree with @vishh that image pulling failures could be transient and retries are needed. Perhaps we can do a better job distinguishing permanent errors such as the image does not exist, and surface them.

**Comment by bgrant0607** [comment]
See also #19343

**Comment by bgrant0607** [comment]
I also agree that image pull failures could be due to races and transient failures.

I like the idea of surfacing this information to the user through status and/or events.

We intend "kubectl rollout status" to report rollout status updates as they occur. This could be one type of update reported.

**Comment by vishh** [comment]
@bgrant0607: Should the error be exposed via a new `ContainerState`?

**Comment by bgrant0607** [comment]
@vishh No. It's a Reason for ContainerStateWaiting.

**Comment by vishh** [comment]
Is `Reason` string considered a stable API for users to build on top of?

**Comment by bgrant0607** [comment]
I view this as similar to #2529. The system should degrade gracefully in the presence of failures.

We're not continuing to work on kubectl rolling-update, but we are working to address this type of thing in Deployment. 

Currently we stop progressing the rollout when we hit maxUnavailable. That allows the user to configure the desired amount of parallelism and risk.

If a container/pod simply fails to become ready, we can't currently distinguish that from normal behavior, but the idea of "permanently failed" deployments will add a progress timeout that could be used to detect this: #19343.

That doesn't enable fast failure in the case of observed problems, admittedly.

Since it's hard to distinguish permanent from transient failures, we'd probably need to add some kind of flag(s) to request fast failure in the case of certain problems (in addition to the timeout).

Failure to pull the image isn't the only failure scenario. For instance, crash loops and deadlocks of newly updated images or configurations (env, command, args, secrets, configmaps, etc.) are also common.

@vishh So far, I've resisted treating Reason as a stable API. At minimum, I reserve the right to create new, more specific Reason values at any time. If we need to, we could add image-related failure info to ContainerStatus, much as we've added ExitCode, RestartCount, etc. It may be a little tricky to represent due to all the possible means of getting images onto the node (external copy, pull, mount, peer-to-peer copy, etc.). Also, sufficiently slow progress might as well be considered failure.

**Comment by bgrant0607** [comment]
Infant mortality was also discussed a little here: https://github.com/kubernetes/kubernetes/pull/19343#issuecomment-216389868

**Comment by kargakis** [comment]
@smarterclayton mentioned somewhere else that ImagePull errors may be transient.

**Comment by lukeschlather** [comment]
Is there an equivalent issue for kubectl rollout with deployments? They seem to have the same behavior of basically surfacing no useful information when a pod fails to start for any reason.
